zip-0227.rst

```diff
@@ -9,7 +9,445 @@
+- For each recipient :math:`i`:
+
+    - generate a ZSA output note that includes the Asset Base. For an Orchard-based ZSA note this is :math:`\mathsf{note}_i = (\mathsf{d}_i, \mathsf{pk}_{\mathsf{d},i}, \mathsf{v}_i, \rho_i, \psi_i, \mathsf{AssetBase^{Orchard}}, \mathsf{rcm}_i)\!`.
+
+- encode the `IssueAction` into the vector `vIssueActions` of the bundle
+
+For the `IssueBundle`:
+
+- encode the `vIssueActions` vector
+- encode the :math:`\mathsf{ik}` as 32 byte-string
+- sign the `SIGHASH` of the transaction with the issuance authorizing key, :math:`\mathsf{isk}`, using the :math:`\mathsf{IssueAuthSig}` signature scheme. The signature is then added to the issuance bundle.
+
+
+NOTE that the commitment is not included in the `IssuanceAction` itself. As explained below, it is later computed by the validators and added to the `NoteCommitmentTree`.
+
+
+Consensus Rule Changes
+----------------------
+
+For the IssueBundle:
+
+- Verify the RedPallas-based issuance authorization signature on `SIGHASH`, `issueAuthSig`, is verified by invoking `issueAuthSig.VerifySig(ik, SIGHASH)`
+
+For each `IssueAction` in `IssueBundle`:
+
+- check that :math:`0 < \mathsf{assetDescSize} <= 512`.
+- check that :math:`\mathsf{asset\_desc}` is a string of length :math:`\mathsf{assetDescSize}` bytes.
+- retrieve :math:`\mathsf{AssetBase}` from the first note in the sequence and check that :math:`\mathsf{AssetBase}` is derived from the issuance validating key :math:`\mathsf{ik}` and :math:`\mathsf{asset\_desc}` as described in the `Asset Identifier`_ section.
+- check that the :math:`\mathsf{AssetId}` does not exist in the ``previously_finalized`` set in the global state.
+- check that every note in the `IssueAction` contains the same :math:`\mathsf{AssetBase}` and is properly constructed as :math:`note = (\mathsf{g_d, pk_d, v, \rho, \psi, AssetBase})`.
+
+If all of the above checks pass, do the following:
+
+- For each note, compute the note commitment as :math:`\mathsf{cm} = \mathsf{NoteCommit^{OrchardZSA}_{rcm}(repr_{\mathbb{P}}(g_d), repr_{\mathbb{P}}(pk_d), v, \rho, \psi, AssetBase)}` as defined in the Note Structure and Commitment section of ZIP 226 [#zip-0226-notestructure]_ and
+- add :math:`\mathsf{cm}` to the Merkle tree of note commitments, `NoteCommitmentTree`.
+- If ``finalize = 1``, add :math:`\mathsf{AssetId}` to the ``previously_finalized`` set immediately after the block in which this transaction occurs.
+
+
+Rationale
+=========
+The following is a list of rationale for different decisions made in the proposal:
+
+- The issuance key structure is independent of the original key tree, but derived in an analogous manner (via ZIP 32). This is in order to keep the issuance details and the Asset Identifiers consistent across multiple shielded pools.
+- The design decision is not to have a chosen name to describe the Custom Asset, but to delegate it to an off-chain mapping, as this would imply a land-grab “war”.
```

Comment:
Wallet developers are probably going to be tempted to display just the `asset_desc` string to their users as the name for an asset, because it's the easy and obvious thing to do. That would be a security issue, because anyone could issue another asset with the same `asset_desc` and confuse those wallets' users into thinking they received the original asset.

Enforcing globally-unique `asset_desc`s would solve that, but then we would have to consider ideograph attacks, which we probably don't want to get into.

I think the ZIP should include something like "Wallets MUST NOT display `asset_desc` to their users as the name of an asset type." And maybe some other recommendations for how wallets should safely map AssetIDs to human-readable names.

It might be convenient to define an asset fingerprint, so that a bridge or whatever can say "Our wrapped BTC asset is `A17294418BD30DEFA1E229658DAE9D7D`" or even better if a user can scan that information into their wallet using a QR code.

Reply:
We've added this recommendation in [this PR](https://github.com/QED-it/zips/pull/32)